torch/_inductor/bounds.py

```diff
@@ -22,12 +22,22 @@ class BoundVars:
 
     def __init__(self, loop_body: LoopBody):
         self.loop_body = loop_body
-        self.replacement_vals = {
-            k: ValueRanges(0, v - 1)
-            if not free_symbols(v)
-            else ValueRanges(2, math.inf)
-            for k, v in loop_body.var_ranges.items()
-        }
+        # Variable substitutions for any constrained ranges:
+        subs = []
+        tracing_context = torch._guards.TracingContext.get()
+        if tracing_context:
+            subs = [
+                (k, v.upper)
+                for k, v in tracing_context.fake_mode.shape_env.var_to_range.items()
+            ]
+        self.replacement_vals = {}
+        for k, v in loop_body.var_ranges.items():
+            s = v.subs(subs)
```

Comment:
Is `var_ranges` just a symint, or can it be a expression? If it can be an expression, you should use `bound_sympy` to compute the actual bounds given the `subs`. If it's not, we can simply look up `v` in `subs` and get its range if it has `free_symbols`.

Reply:
@lezcano (Pardon my being a noob here). `loop_body.var_ranges` can contain expressions, e.g., I see `4 * s0`. Given that, are you simply suggesting to use the `bound_sympy` helper to evaluate the bounds instead of the `subs` method I've used here? Playing around with that helper function, it seems like it indeed probably does what I want. But it sounds like you anticipate some modifications are needed in that helper first? If so, I don't understand that part.

Reply:
Have a look at how `bound_sympy` is implemented. Before calling into the bulk of the logic, it tries to do something very similar to what you're doing here. Just move the logic you implemented here to `bound_sympy` and you should be able to avoid repeating it :)

Reply:
Thanks. I might be naive, but looks like `bound_sympy` might "just work". I'll put up a new rev to show you what I mean.